### PR TITLE
[libc] Add `dlfcn.h` placeholder

### PR DESCRIPTION
**Pull request by izaakschroeder**
Adds `dlopen` and friends. This is needed as part of the effort to compile `libunwind` + `libc` without baremetal mode. This is part of https://github.com/llvm/llvm-project/issues/97191. This should still be spec compliant, since `dlopen` always returns `NULL` and `dlerror` always returns an error message.

> If dlopen() fails for any reason, it returns NULL.

> The function dlclose() returns 0 on success, and nonzero on error.

> Since the value of the symbol could actually be NULL (so that a NULL return from dlsym() need not indicate an error), the correct way to test for an error is to call dlerror() to clear any old error conditions, then call dlsym(), and then call dlerror() again, saving its return value into a variable, and check whether this saved value is not NULL.

See:
- https://linux.die.net/man/3/dlopen

### CODE REVIEW

File: libc/src/dlfcn/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,43 @@
+add_entrypoint_object(
+  dlclose
+  SRCS
+    dlclose.cpp
+  HDRS
+    dlclose.h
+  DEPENDS
+    libc.include.dlfcn
+    libc.src.errno.errno
```
Comment (overmighty) [inline libc/src/dlfcn/CMakeLists.txt:9]: Your current implementations don't include `dlfcn.h` or use `errno`.

```suggestion
```

File: libc/src/dlfcn/dlerror.cpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+//===----------------------------------------------------------------------===//
+
+#include "dlerror.h"
+
+#include "src/__support/common.h"
+
+namespace LIBC_NAMESPACE {
+
+LLVM_LIBC_FUNCTION(char *, dlerror, ()) {
+  return const_cast<char *>("unsupported");
```
Comment (overmighty) [inline libc/src/dlfcn/dlerror.cpp:16]: Not sure if that `const_cast` is fine. Despite the non-const `char *` return type, POSIX does state that "The application shall not modify the string returned": https://pubs.opengroup.org/onlinepubs/9699919799/functions/dlerror.html.

Comment (izaakschroeder) [reply libc/src/dlfcn/dlerror.cpp:16]: I had the same thoughts… it looks like the caller is not supposed to touch this pointer so this seemed safe to me? If someone knows better lmk and I will change it.

**Comment by overmighty** [comment]
There weren't any merge conflicts, were there? If you amend and force-push to have only 1 commit on the pull request, then you shouldn't do that. See https://llvm.org/docs/GitHub.html#updating-pull-requests.

Even if there are merge conflicts, you should rebase your existing commits instead of resetting your branch and creating a single new commit.

**Comment by izaakschroeder** [comment]
> There weren't any merge conflicts, were there? If you amend and force-push to have only 1 commit on the pull request, then you shouldn't do that. See https://llvm.org/docs/GitHub.html#updating-pull-requests.
> 
> Even if there are merge conflicts, you should rebase your existing commits instead of resetting your branch and creating a single new commit.

Ok I can do that no problem 😀 I do like the idea of tracking review changes... though my habit has normally been to clean everything up before merging and I guess I had figured we were pretty close. Thanks for the heads up!

**Review by SchrodingerZhu — COMMENTED** [review]
LGTM. But I would like to put some TODOs inside the comment. Also, please add a comment about the `dlerror` return in the `Undefined Behavior` document section: https://github.com/llvm/llvm-project/blob/main/libc/docs/dev/undefined_behavior.rst

**Comment by lntue** [comment]
> LGTM. But I would like to put some TODOs inside the comment. Also, please add a comment about the `dlerror` return in the `Undefined Behavior` document section: https://github.com/llvm/llvm-project/blob/main/libc/docs/dev/undefined_behavior.rst

+1 on adding TODO.  It would be even better if you don't mind creating github issues for those TODOs and link to them in the comments.

**Review by SchrodingerZhu — APPROVED** [review]
LGTM

**Pull request merged**